src/renderer.rs

```diff
@@ -66,6 +87,46 @@ pub fn fill_default<T: Default>(vec: &mut Vec<T>, size: usize) {
 	vec.resize_with(size, T::default);
 }
 
+/// Query which link (if any) is under the provided mupdf device coordinates for the given
+/// page. Returns Ok(Some(LinkTarget)) if a link is found, Ok(None) if no link, or
+/// Err(mupdf::error::Error) on mupdf failures.
+fn query_link_at(
+	doc: &Document,
+	qpage: usize,
+	mupdf_x_px: f32,
+	mupdf_y_px: f32,
+	area_w: f32,
+	area_h: f32,
+	fit_or_fill: FitOrFill,
+) -> Result<Option<LinkTarget>, mupdf::error::Error> {
+	// load the requested page and compute same scale as used when rendering
+	let page = doc.load_page(qpage as i32)?;
+	let bounds = page.bounds()?;
+	let page_dim = (bounds.x1 - bounds.x0, bounds.y1 - bounds.y0);
+	let scaled = scale_img_for_area(page_dim, (area_w, area_h), fit_or_fill);
+	let scale_factor = scaled.scale_factor;
+
+	let pdf_x = mupdf_x_px / scale_factor;
+	let pdf_y = mupdf_y_px / scale_factor;
+
+	if let Ok(mut links) = page.links() {
+		while let Some(link) = links.next() {
+			let lb = link.bounds;
+			if pdf_x >= lb.x0 && pdf_x <= lb.x1 && pdf_y >= lb.y0 && pdf_y <= lb.y1 {
+				// prefer URI if present and non-empty
+				if !link.uri.is_empty() {
+					return Ok(Some(LinkTarget::Uri(link.uri)));
+				}
+				if let Some(dest) = link.dest {
+					return Ok(Some(LinkTarget::GoTo { page_index: dest.loc.page_number as usize }));
+				}
```

Comment:
This now makes me wonder what we should do if we encounter a link that has no destination and an empty URI - do you think we should report that to the user? Maybe that would just be a useless notification, since they probably don't care about links that effectively aren't links... idk, this is probably fine as-is, just something to think about in case you think something should be done there.